File: R/sparse_gcPCA.R
# sparse_gcPCA.R

# Implementation of sparse gcPCA in R, it is not yet guaranteed that this implementation is correct and works, still need to run some tests
sparse_gcPCA <- function(Ra, Rb, method = 'v4', Ncalc = NULL, normalize_flag = TRUE, Nsparse = NULL, Nshuffle = 0,
                         lasso_penalty = exp(seq(log(1e-2), log(1), length.out = 10)), ridge_penalty = 0,
                         alpha = 1, alpha_null = 0.975, tol = 1e-5, max_steps = 1000, cond_number = 1e13) {
  
  stopifnot(is.matrix(Ra), is.matrix(Rb))
  stopifnot(ncol(Ra) == ncol(Rb))
  stopifnot(all(lasso_penalty >= 0))
  stopifnot(method %in% c("v1","v2","v2.1","v3","v3.1","v4","v4.1"))
  
  if (is.null(Ncalc)) Ncalc <- ncol(Ra)
  if (is.null(Nsparse)) Nsparse <- ncol(Ra)

  # fit base gcPCA model
  gc_model <- gcPCA(Ra, Rb, method, Ncalc, Nshuffle, normalize_flag, alpha, alpha_null, cond_number)
  
  # extract needed components
  Ra <- gc_model$Ra
  Rb <- gc_model$Rb
  Jorig <- gc_model$J

  # Legacy safeguard: some gcPCA versions returned a shrunken J (orthogonal complement
  # after the first loading) instead of the original SVD basis.
  target_gcpcs <- ncol(gc_model$loadings)
  if (is.null(Jorig) || ncol(Jorig) < target_gcpcs) {
    svd_result <- svd(rbind(Ra, Rb))
    Jorig <- svd_result$v[, 1:target_gcpcs, drop = FALSE]
    warning("Recovered original J basis from combined SVD because gc_model$J was reduced.")
  }
  p <- ncol(Ra)
  
  # fitting function
  sparse_fit <- function() {

    # projecting data to shared components
    RaJ <- Ra %*% Jorig
    RbJ <- Rb %*% Jorig

    # covariance matrices
    JRaRaJ <- (t(RaJ) %*% RaJ) / (nrow(RaJ) - 1)
    JRbRbJ <- (t(RbJ) %*% RbJ) / (nrow(RbJ) - 1)
    
    # v1 is equivalent to sparse contrastive PCA
    if (method == 'v1') {

      # Compute theta matrix
      theta <- JRaRaJ - alpha * JRbRbJ
      eig <- eigen(theta, symmetric = TRUE)
      w <- eig$values
      v <- eig$vectors
      
      # Split positive/negative eigenvalues and allocate sparse dimensions
      new_w_pos <- pmax(w, 0)
      new_w_neg <- -pmin(w, 0)

      n_target <- if (is.infinite(Nsparse)) length(w) else min(as.integer(Nsparse), length(w))
      n_gcpcs_pos <- min(sum(w > 0), n_target)
      n_gcpcs_neg <- min(sum(w < 0), n_target - n_gcpcs_pos)

      theta_pos <- v %*% diag(sqrt(new_w_pos)) %*% t(v)
      theta_neg <- v %*% diag(sqrt(new_w_neg)) %*% t(v)

      final_loadings <- lapply(lasso_penalty, function(lmbda) {
        pos_load <- if (n_gcpcs_pos > 0) {
          J_variable_projection(theta_pos, Jorig, n_gcpcs_pos, lmbda, ridge_penalty, max_steps, tol)
        } else {
          matrix(numeric(0), nrow = nrow(Jorig), ncol = 0)
        }

        neg_load <- if (n_gcpcs_neg > 0) {
          J_variable_projection(theta_neg, Jorig, n_gcpcs_neg, lmbda, ridge_penalty, max_steps, tol)
        } else {
          matrix(numeric(0), nrow = nrow(Jorig), ncol = 0)
        }

        cbind(pos_load, neg_load)
      })
      
    } else {
      # v2-v4 methods
      if (method %in% c('v2', 'v2.1')) {
        numerator <- JRaRaJ
        denominator <- JRbRbJ
      } else if (method %in% c('v3', 'v3.1')) {
        numerator <- JRaRaJ - JRbRbJ
        denominator <- JRbRbJ
      } else if (method %in% c('v4', 'v4.1')) {
        numerator <- JRaRaJ - JRbRbJ
        denominator <- JRaRaJ + JRbRbJ
      }
      
      # find M matrix
      eig_denom <- eigen(denominator, symmetric = TRUE)
      M <- eig_denom$vectors %*% diag(sqrt(eig_denom$values)) %*% t(eig_denom$vectors)
      Minv <- solve(M)
      sigma <- t(Minv) %*% numerator %*% Minv

      eig_sigma <- eigen(sigma, symmetric = TRUE)
      w <- eig_sigma$values
      v <- eig_sigma$vectors

      # all positive eigenvalues
      new_w <- w + 2

      theta_pos <- v %*% diag(sqrt(new_w)) %*% t(v)
      
      # sparse loading
      final_loadings <- lapply(lasso_penalty, function(lmbda) {
        J_M_variable_projection(theta_pos, Jorig, M, Nsparse, lmbda, ridge_penalty, max_steps, tol)
      })
    }
    
    # normalize loading
    sparse_loadings <- lapply(final_loadings, function(load) {
      sweep(load, 2, apply(load, 2, function(x) norm(x, "2")), "/")
    })
    
    Ra_scores <- lapply(final_loadings, function(L) {
      temp <- Ra %*% L                 # projecting
      temp_norm <- apply(temp, 2, norm, type = "2") # getting the norm
      temp_norm[temp_norm == 0] <- 1   # avoid division by zero
      sweep(temp, 2, temp_norm, "/")   # norm 1 scores
    })
    
    Ra_values <- lapply(final_loadings, function(L) {
      temp <- Ra %*% L
      apply(temp, 2, norm, type = "2") # same as LA.norm(temp, axis=0)
    })

    Rb_scores <- lapply(final_loadings, function(L) {
      temp <- Rb %*% L                 # projecting
      temp_norm <- apply(temp, 2, norm, type = "2") # getting the norm
      temp_norm[temp_norm == 0] <- 1   # avoid division by zero
      sweep(temp, 2, temp_norm, "/")   # norm 1 scores
    })

    Rb_values <- lapply(final_loadings, function(L) {
      temp <- Rb %*% L
      apply(temp, 2, norm, type = "2") # same as LA.norm(temp, axis=0)
    })

    return(list(sparse_loadings = sparse_loadings, original_loadings = gc_model$loadings, Ra_scores = Ra_scores, Ra_values = Ra_values,
                Rb_scores = Rb_scores, Rb_values = Rb_values, Ra = Ra,
                Rb = Rb, J = Jorig, normalize_flag = normalize_flag, gcPCA_object = gc_model))
  }
  
  l2_norm_vec <- function(x) {
    sqrt(sum(x^2))
  }
  
  # J and M variable projection
  J_M_variable_projection <- function(theta_input, J, M, k, alpha, beta, max_iter, tol) {
    svd_theta <- svd(theta_input)
    Dmax <- svd_theta$d[1]
    B <- svd_theta$v[, 1:k, drop = FALSE]
    
    VD <- sweep(svd_theta$v, 2, svd_theta$d, "*")
    VD2 = sweep(svd_theta$v, 2, svd_theta$d^2, "*")
    
    # tuning parameters
    alpha_scaled <- alpha * Dmax^2
    beta_scaled <- beta * Dmax^2
    nu <- 1 / (Dmax^2 + beta_scaled)
    kappa <- nu * alpha_scaled
    
    obj <- numeric(0)
    improvement <- Inf
    
    # reducing computations in the loop
    VD2_Vt = VD2 %*% t(svd_theta$v)
    Minv <- solve(M)
    JMinv <- J %*% Minv
    MJt <- M %*% t(J)
    
    for (iter in 1:max_iter) {
      Z <- VD2_Vt %*% B
      svd_Z <- svd(Z)
      A <- svd_Z$u %*% t(svd_Z$v)
      
      # gradient update
      grad <- (VD2_Vt %*% (A - B)) - beta_scaled * B
      B_temp <- JMinv %*% B + nu * JMinv %*% grad
      
      # L1 soft thresholding
      Bf <- ifelse(B_temp > kappa, B_temp - kappa,
                   ifelse(B_temp < -kappa, B_temp + kappa, 0))
      
      # returning it to the Y space from the feature space
      B <- MJt %*% Bf
      
      R = t(VD) - t(VD) %*% B %*% t(A)
      obj_value = 0.5 * sum(R^2) + alpha_scaled * sum(abs(B)) + 0.5 * beta_scaled * sum(B^2)
      obj <- c(obj, obj_value)
      
      # check convergence
      if (iter > 1) {
        improvement <- (obj[iter-1] - obj[iter]) / obj[iter]
        if (improvement < tol) break
      }
    }
    # l2 norm for normalization
    l2_norms <- apply(Bf, 2, l2_norm_vec)
    
    # loadings normalization
    loadings_ = sweep(Bf, 2, l2_norms, FUN = "/")
    
    return(loadings_)
  }
  
  # J variable projection (used for method v1)
  J_variable_projection <- function(theta_input, J, k, alpha, beta, max_iter, tol) {
    svd_theta <- svd(theta_input)
    Dmax <- svd_theta$d[1]
    k_eff <- min(k, ncol(svd_theta$v))
    B <- svd_theta$v[, 1:k_eff, drop = FALSE]

    VD <- sweep(svd_theta$v, 2, svd_theta$d, "*")
    VD2 <- sweep(svd_theta$v, 2, svd_theta$d^2, "*")

    # tuning parameters
    alpha_scaled <- alpha * Dmax^2
    beta_scaled <- beta * Dmax^2
    nu <- 1 / (Dmax^2 + beta_scaled)
    kappa <- nu * alpha_scaled

    obj <- numeric(0)
    improvement <- Inf
    VD2_Vt <- VD2 %*% t(svd_theta$v)
    Bf <- B

    for (iter in 1:max_iter) {
      # update A
      Z <- VD2_Vt %*% B
      svd_Z <- svd(Z)
      A <- svd_Z$u %*% t(svd_Z$v)

      # gradient update in Y-space and map to feature-space through J
      grad <- (VD2_Vt %*% (A - B)) - beta_scaled * B
      B_temp <- J %*% B + nu * J %*% grad

      # l1 soft thresholding
      Bf <- ifelse(B_temp > kappa, B_temp - kappa,
                   ifelse(B_temp < -kappa, B_temp + kappa, 0))

      # map back to Y-space
      B <- t(J) %*% Bf

      R <- t(VD) - t(VD) %*% B %*% t(A)
      obj_value <- 0.5 * sum(R^2) + alpha_scaled * sum(abs(B)) + 0.5 * beta_scaled * sum(B^2)
      obj <- c(obj, obj_value)

      if (iter > 1) {
        improvement <- (obj[iter - 1] - obj[iter]) / obj[iter]
        if (improvement < tol) break
      }
    }

    # normalize loadings per column
    l2_norms <- apply(Bf, 2, l2_norm_vec)
    l2_norms[l2_norms == 0] <- 1
    loadings_ <- sweep(Bf, 2, l2_norms, FUN = "/")
    return(loadings_)
  }
  
  # execute and return results
  result <- sparse_fit()
  class(result) <- "sparse_gcPCA"
  return(result)
}

predict.sparse_gcPCA <- function(object, Ra = NULL, Rb = NULL, ...) {
  # object: result of sparse_gcPCA
  # Ra/Rb: matrices with same columns as the Ra/Rb used in fitting
  #
  # Returns:
  #   - If Ra and Rb are NULL: the training scores/values stored in the object
  #   - Otherwise: updated scores/values for any supplied Ra and/or Rb
  #     (still as lists, one entry per lasso penalty)

  # If no new data, just return what was fitted
  if (is.null(Ra) && is.null(Rb)) {
    Ra_scores <- lapply(seq_along(object$Ra_scores), function(i) {
        scores <- object$Ra_scores[[i]]  # n x k_i
        vals   <- object$Ra_values[[i]]  # length k_i
        sweep(scores, 2, vals, "*")      # n x k_i, raw transform
    })
    
    Rb_scores <- lapply(seq_along(object$Rb_scores), function(i) {
        scores <- object$Rb_scores[[i]]  # n x k_i
        vals   <- object$Rb_values[[i]]  # length k_i
        sweep(scores, 2, vals, "*")      # n x k_i, raw transform
      })

    return(list(
      Ra_scores = Ra_scores,
      Rb_scores = Rb_scores
    ))
  }

  # Initialize
  Ra_scores <- NULL
  Rb_scores <- NULL

  # ancillary function to project data onto a list of sparse loading matrices
  project_sparse <- function(X, loadings_list) {
    X <- as.matrix(X)

    scores_list <- vector("list", length(loadings_list))
    # values_list <- vector("list", length(loadings_list))

    for (i in seq_along(loadings_list)) {
      L <- loadings_list[[i]]              # p x k_i
      temp_scores <- X %*% L                      # n x k_i

      scores_list[[i]] <- temp_scores
      # Norm-1 scores (same as in sparse_gcPCA fitting)
      # scores_list[[i]] <- sweep(temp, 2, norms, "/")
      # values_list[[i]] <- norms
    }

    list(scores = scores_list)
  }

  sparse_loadings <- object$sparse_loadings
  # normalize_flag <- isTRUE(object$normalize_flag)

  # Try to pull the same normalization function used in gcPCA()
  # normalize_fun <- NULL
  # if (!is.null(object$gcPCA_object) &&
  #     !is.null(object$gcPCA_object$normalize)) {
  #   normalize_fun <- object$gcPCA_object$normalize
  # }

  # Process Ra if provided
  if (!is.null(Ra)) {
    Ra_proc <- as.matrix(Ra)


    # if (normalize_flag && !is.null(normalize_fun)) {
      # Ra_proc <- normalize_fun(Ra_proc)
    # }

    proj_Ra <- project_sparse(Ra_proc, sparse_loadings)
    Ra_scores <- proj_Ra$scores
  }

  # Process Rb if provided
  if (!is.null(Rb)) {
    Rb_proc <- as.matrix(Rb)

    # if (normalize_flag && !is.null(normalize_fun)) {
    #   Rb_proc <- normalize_fun(Rb_proc)
    # }

    proj_Rb <- project_sparse(Rb_proc, sparse_loadings)
    Rb_scores <- proj_Rb$scores
  }

  return(list(
    Ra_scores = Ra_scores,
    Rb_scores = Rb_scores
  ))
}

File: matlab/sparse_gcPCA.m
function [B, S, X] = sparse_gcPCA(Ra, Rb, gcPCAversion, varargin)
% function [B, S, X] = sparse_gcPCA(Ra, Rb, gcPCAversion, varargin)
%
% This function performs sparse generalized contrastive PCA (gcPCA), which takes
% the input matrices Ra and Rb and finds the dimensions that differ most
% between them with penalty terms to create sparseness in the loadings.
% gcPCA is analogous to PCA/SVD, which finds dimensions that
% maximize variance in the data matrix Z, decomposing it into (U * S * V'),
% where U contains scores, V contains loadings, and S is a diagonal matrix
% showing how much variance each PC accounts for. Instead, gcPCA decomposes
% Ra and Rb into (Ba * Sa * X') and (Bb * Sb * X'), respectively. Unlike in
% PCA, Ba, Bb, and X are not generally orthogonal
%
% INPUTS
% Ra             -- (p1 x N) matrix of data (N features, p1 datapoints)
% Rb             -- (p2 x N) matrix of data (N features, p2 datapoints)
% gcPCAversion   -- version of gcPCA to use
% Nshuffle       -- (optional) number of times to shuffle for null
%                   distribution of S
% Ncalc          -- (optional) maximum number of gcPCs to calculate (useful
%                   only for orthogonal gcPCA, which is iterative)
% normalize      -- (optional) logical variable to normalize the data or
%                   not, default true. Remeber to at least center your data
%                   if turning this off.
%
% OUTPUTS
% B (struct)     -- gcPCA scores (different for Ra and Rb)
% S (struct)     -- amount of variance accounted for by each gcPC
% X              -- gcPCA loadings (shared between Ra and Rb)
%
% There are currently six versions of gcPCA, which find
% dimensions maximizing different objective functions:
%
% v2.0: maximizes Ra/Rb, X not constrained to be orthogonal
% v2.1: maximizes Ra/Rb, X constrained to be orthogonal
%
% v3.0: maximizes (Ra+Rb)/Rb, X not constrained to be orthogonal
% v3.1: maximizes (Ra+Rb)/Rb, X constrained to be orthogonal
%
% v4.0: maximizes (Ra+Rb)/(Ra+Rb), X not constrained to be orthogonal
% v4.1: maximizes (Ra+Rb)/(Ra+Rb), X constrained to be orthogonal
%
%
% Eliezyer de Oliveira, 2025-01-07



%% checking parameters and input arguments
pars = inputParser;
pars.addRequired('Ra');
pars.addRequired('Rb');
pars.addRequired('gcPCAversion');
pars.addOptional('Nshuffle', 0, @isnumeric);
pars.addOptional('normalize', true, @islogical);
pars.addOptional('Nsparse', 2, @isnumeric);
pars.addOptional('lasso_penalty', exp(linspace(log(1e-2),log(1),10)), @isnumeric);
pars.addOptional('ridge_penalty', 0, @isnumeric);
pars.addOptional('alpha', 1, @isnumeric);
pars.addOptional('maxcond', 10^13, @isnumeric); % you shouldn't need to change this, but it's condition number to regularize the denominator matrix if it's ill-conditioned
pars.addOptional('maxiter', 1000, @isnumeric);
pars.addOptional('tol', 1e-5, @isnumeric);

pars.parse('Ra', 'Rb', 'gcPCAversion', varargin{:});

maxcond = pars.Results.maxcond;
Nsparse = pars.Results.Nsparse;
alpha = pars.Results.alpha;
Nshuffle = pars.Results.Nshuffle;
normalize_flag = pars.Results.normalize;
lasso_penalty = pars.Results.lasso_penalty;
ridge_penalty = pars.Results.ridge_penalty;
alpha = pars.Results.alpha;
maxiter = pars.Results.maxiter;
tol = pars.Results.tol;

%% Step 1: normalize inputs if necessary
N = size(Rb, 2); % number of dimensions
if size(Ra, 2) ~= N
    error('Ra and Rb have different numbers of dimensions');
end

if normalize_flag
    Rb_temp = normalize(zscore(Rb), 'norm');
    Ra_temp = normalize(zscore(Ra), 'norm');
    
    if sum((Rb_temp(:) - Rb(:)) .^2) > (0.01 .* Rb_temp.^2)
        warning('Rb was not normalized properly - normalizing now');
        Rb = Rb_temp;
    end
    if sum((Ra_temp(:) - Ra(:)) .^2) > (0.01 .* Ra_temp.^2)
        warning('Ra was not normalized properly - normalizing now');
        Ra = Ra_temp;
    end
    clear Rb_temp Ra_temp
end

%% Step 2: do SVD and discard dimensions if necessary
p = min(size(Ra, 1), size(Rb, 1)); % we use the p from whichever dataset has fewer datapoints
N_gcPCs = min(p, N); % number of gcPCs to calculate. If p < N (meaning
% fewer datapoints than dimensions), we will only calculate p gcPCs

% doing SVD on the combined dataset. The goal here is to discard dimensions
% that have near-zero variance in *both* Ra and Rb
RaRb = [Ra; Rb];
tol = max(size(RaRb)) * eps(norm(RaRb)); % default cutoff used by rank()
[~, Sab, J] = svd(RaRb, 'econ'); % calculate SVD on combined data

Sab = diag(Sab);
if sum(Sab>tol) < N_gcPCs % data is rank-deficient
    warning('Input data is rank-deficient! Discarding dimensions; cannot shuffle.');
    N_gcPCs = sum(Sab>tol);
    Nshuffle = 0; % can't shuffle because it's not a valid null
end

% number of dimensions to return is either the amount requested by user or
% the amount of available dimensions, whichever is smaller
% N_gcPCs = min(N_gcPCs, Nsparse);

Jorig = J(:, 1:N_gcPCs); % using the first N_gcPCs dimensions
J = Jorig; % J shrinks every round, but Jorig is the first-round's J
clear RaRb Sab


%% Step 3: solving sparse gcPCA

if gcPCAversion == 1
    RaJ = Ra * J; % projecting into lower-D subspace spanned by J
    RbJ = Rb * J;
    
    JRaRaJ = RaJ'*RaJ / (size(RaJ,1)-1);
    JRbRbJ = RbJ'*RbJ / (size(RbJ,1)-1);
    obj_info = 'Ra - alpha*Rb';
    clear RaJ RbJ
    
    sigma = JRaRaJ - alpha*JRbRbJ;
    [y, D] = eig(sigma);
    D = diag(D);
    
    % Calculating only the number of dimensions requested by user
    n_gcpcs_pos = sum(D > 0);
    if (n_gcpcs_pos - Nsparse) >= 0
        n_gcpcs_pos = Nsparse;
        n_gcpcs_neg = 0;
    else
        temp = Nsparse - n_gcpcs_pos;
        n_gcpcs_neg = temp;
    end
    
    % Separating positive and negative eigenvalues
    Dpos = D;
    Dneg = D;
    
    %separating positive and negative eigenvalues
    Dpos(Dpos<0) = 0;
    Dneg(Dneg>0) = 0;
    
    % Square root matrix of the positive and 'negative' eigenvalues
    alpha_pos = max(Dpos)/maxcond;  % fixing the condition number
    theta_pos = y * diag(sqrt(Dpos + alpha_pos)) * y';
    
    alpha_neg = max(Dneg)/maxcond;  % fixing the condition number
    theta_neg = y * diag(sqrt(-Dneg + alpha_neg)) * y';
    
    % if there is any positive eigenvalue
    if n_gcpcs_pos>0
        final_pos_loadings = [];
        for lambda_idx = 1:length(lasso_penalty)
            lambda = lasso_penalty(lambda_idx);
            Bf = J_variable_projection(theta_pos, J, 'k', n_gcpcs_pos, 'alpha', lambda, 'beta', ridge_penalty, 'maxiter', maxiter, 'tol', tol);
            final_pos_loadings = [final_pos_loadings, normalize(Bf, 'norm')];
        end
        
    else
        final_pos_loadings = [];
    end
    
    % if there is any negative eigenvalue
    if n_gcpcs_neg>0
        final_neg_loadings = [];
        for lambda_idx = 1:length(lasso_penalty)
            lambda = lasso_penalty(lambda_idx);
            Bf = J_variable_projection(theta_neg, J, 'k', n_gcpcs_neg, 'alpha', lambda, 'beta', ridge_penalty, 'maxiter', maxiter, 'tol', tol);
            final_neg_loadings = [final_neg_loadings, normalize(Bf, 'norm')];
        end
        
    else
        final_neg_loadings = [];
    end
    
    % rearranging the PCs by vector and concatenating
    final_loadings = [];
    for lambda_idx = 1:length(lasso_penalty)
        if n_gcpcs_pos>0 && n_gcpcs_neg>0
            array_idx = Nsparse*(lambda_idx-1)+1:Nsparse*lambda_idx;
            final_loadings = [final_loadings , cat(2, final_pos_loadings(:,array_idx), final_neg_loadings(:,array_idx))];
        elseif n_gcpcs_pos == 0 && n_gcpcs_neg>0
            array_idx = Nsparse*(lambda_idx-1)+1:Nsparse*lambda_idx;
            final_loadings = [final_loadings, final_neg_loadings(:,array_idx)];
        else
            array_idx = Nsparse*(lambda_idx-1)+1:Nsparse*lambda_idx;
            final_loadings = [final_loadings, final_pos_loadings(:,array_idx)];
        end
        
    end
    
else
    
    % calculate numerator and denominator for objective function
    RaJ = Ra * J; % projecting into lower-D subspace spanned by J
    RbJ = Rb * J;
    
    JRaRaJ = RaJ'*RaJ / (size(RaJ,1)-1);
    JRbRbJ = RbJ'*RbJ / (size(RbJ,1)-1);
    if gcPCAversion == 2 % calculating gcPCA using Ra/Rb objective function
        numerator = JRaRaJ;
        denominator = JRbRbJ;
        obj_info = 'Ra ./ Rb';
        
    elseif gcPCAversion == 3 % calculating gcPCA using (Ra-Rb)/Rb objective function
        numerator = JRaRaJ - JRbRbJ;
        denominator = JRbRbJ;
        obj_info = '(Ra-Rb) ./ Rb';
        
    elseif gcPCAversion == 4 % calculating gcPCA using (Ra-Rb)/(Ra+Rb) objective function
        numerator = JRaRaJ - JRbRbJ;
        denominator = JRaRaJ + JRbRbJ;
        obj_info = '(Ra-Rb) ./ (Ra+Rb)';
    end
    
    %  Define numerator and denominator according to the method requested
    if gcPCAversion == 2
        numerator = JRaRaJ;
        denominator = JRbRbJ;
        obj_info = 'Ra / Rb';
    elseif gcPCAversion == 3
        numerator = JRaRaJ - JRbRbJ;
        denominator = JRbRbJ;
        obj_info = '(Ra-Rb) / Rb';
    elseif gcPCAversion == 4
        numerator = JRaRaJ - JRbRbJ;
        denominator = JRaRaJ + JRbRbJ;
        obj_info = '(Ra-Rb) / (Ra+Rb)';
    else
        error('Version input not recognized, please pick between v1-v4')
    end
    
    % getting the square root matrix of denominator
    [e, d] = eig(denominator);
    M = e * sqrt(d) * e';  % getting square root matrix M
    Minv = inv(M);
    
    sigma = Minv'*numerator*Minv;
    
    %
    % Getting square root matrix of sigma
    [v,w] = eig(sigma);
    
    
    % off setting the eigenvalues to be positive definite
    new_w = diag(diag(w) + 2);  % adding 2 to make it positive definite
    theta_pos = v * sqrt(new_w) * v';
    
    final_loadings = [];
    for lambda_idx = 1:length(lasso_penalty)
        lambda = lasso_penalty(lambda_idx);
        Bf = J_M_variable_projection(theta_pos, J, M, 'k', Nsparse, 'alpha', lambda, 'beta', ridge_penalty, 'maxiter', maxiter, 'tol', tol);
        final_loadings = [final_loadings, normalize(Bf, 'norm')];
    end
end

sparse_loadings=reshape(final_loadings,[size(final_loadings,1),Nsparse,length(lasso_penalty)]);
[~,n,p] = size(sparse_loadings); % getting loadings matrix dimensions

% calculating scores and values for condition A
temp_ra_scores = [];
temp_ra_values = [];
for sload_idx = 1:p
    sload = sparse_loadings(:,:,sload_idx);
    temp = Ra*sload;
    temp_norm = vecnorm(temp);
    temp_norm(temp_norm == 0) = 1;
    temp_ra_scores = [temp_ra_scores, temp./temp_norm];
    temp_ra_values = [temp_ra_values, vecnorm(temp)];
end

Ra_scores_ = reshape(temp_ra_scores,size(temp_ra_scores,1),n,p);
Ra_values_ = reshape(temp_ra_values,n,p);

% calculating scores and values for condition B
temp_rb_scores = [];
temp_rb_values = [];
for sload_idx = 1:p
    sload = sparse_loadings(:,:,sload_idx);
    temp = Rb*sload;
    temp_norm = vecnorm(temp);
    temp_norm(temp_norm == 0) = 1;
    temp_rb_scores = [temp_rb_scores, temp./temp_norm];
    temp_rb_values = [temp_rb_values, vecnorm(temp)];
end

Rb_scores_ = reshape(temp_rb_scores,[size(temp_rb_scores,1),n,p]);
Rb_values_ = reshape(temp_rb_values,n,p);

% organize output
B.gcPCAversion = gcPCAversion;
for idx = 1:p
    B.a{idx} = squeeze(Ra_scores_(:,:,idx));
end
B.a_info = 'Scores for Ra on gcPCA, each cell is a lasso penalty parameter';
for idx = 1:p
    B.b{idx} = squeeze(Rb_scores_(:,:,idx));
end
B.b_info = 'Scores for Rb on gcPCA, each cell is a lasso penalty parameter';

S.gcPCAversion = gcPCAversion;
S.objective = obj_info;
for idx = 1:p
    S.a{idx} = Ra_values_(:,idx);
end
S.a_info = 'gcPCA values for Ra, each cell is a lasso penalty parameter';
for idx = 1:p
    S.b{idx} = Rb_values_(:,idx);
end
S.b_info = 'gcPCA values for Rb, each cell is a lasso penalty parameter';
S.lasso_penalty = lasso_penalty;
S.ridge_penalty = ridge_penalty;

for idx = 1:p
    X{idx} = squeeze(sparse_loadings(:,:,idx));
end

end

